### PR TITLE
dynamic routes

### PR DESCRIPTION
**Pull request by jonatjano**
**Summary**

This PR allow dynamic loading of routes

Explain the **motivation** for making this change. What existing problem does the pull request solve?

With this PR you don't need to modify routes.php each time you add a new controller

**Test plan (required)**

I tested using the command from `.travis.yml` : 
```bash
docker run -v "`pwd`:/home/builder/" tracesoftware/gitlab-builder:php7-cli /bin/sh -c "mkdir -p $ARCHIVE_PATH $DOCS_PATH; chmod -R 777 $DATA_PATH; composer update; composer test"
```
and got this result stating everything work : 
capture d ecran de 2018-05-25 11-10-53

EDIT: picture didn't show the first line

### CODE REVIEW

File: BackEnd/src/routes.php [hunk 1]
```diff
@@ -1,15 +1,76 @@
...
-$slim->get('/listProjects', ListProjects::class);
-$slim->post('/addProject', AddProject::class);
-$slim->delete('/deleteProject', DeleteProject::class);
+// declare function used to dynamicaly load routes
+// they are first because since they are conditionnaly declared they have to be first to be recognized
+// they are conditionnaly declared because PHPUnit require the file multiple time
+//      and was going "Fatal error function is already defined"
+
+// get every classes declared in the file
+if(!function_exists("file_get_php_classes")) {
```
Comment (MattMattV) [inline BackEnd/src/routes.php]: No need to perform this check, we will run at least PHP 7

File: BackEnd/src/routes.php [hunk 2]
```diff
@@ -1,15 +1,76 @@
...
+    function file_get_php_classes(string $filepath) {
```
Comment (MattMattV) [inline BackEnd/src/routes.php]: no need to declare this method

File: BackEnd/src/routes.php [hunk 3]
```diff
@@ -1,15 +1,76 @@
...
+        $php_code = file_get_contents($filepath);
+        $classes = get_php_classes($php_code);
+        return $classes;
+    }
+}
+
+// get every classes declared in a valid php string
+if(!function_exists("get_php_classes")) {
```
Comment (MattMattV) [inline BackEnd/src/routes.php]: No need to perform this check, we will run at least PHP 7

**Comment by jonatjano** [comment]
last ones do some big changed so here new tests result : 
capture d ecran de 2018-05-28 15-14-09

for coverage the results are weird and I wanted to know if you saw the reason for this result : 
capture d ecran de 2018-05-28 15-11-54

we see that controller is pretty badly covered but the reason is : 
capture d ecran de 2018-05-28 15-12-00

that function is marked as partly covered while it IS fully covered otherwise tests about getProjects would fail (plus the coverage stop partway without explication :cry: )

**Comment by jonatjano** [comment]
found the issue about coverage, since there was no project the finder stopped...
here the new report ;)
capture d ecran de 2018-05-29 10-20-35

the tests are still succesful

File: BackEnd/src/controllers/ProjectController.php [hunk 4]
```diff
@@ -0,0 +1,234 @@
+<?php
+namespace HostMyDocs\Controllers;
+
+use HostMyDocs\Models\Language;
+use HostMyDocs\Models\Project;
+use HostMyDocs\Models\Version;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\SplFileInfo;
+
+abstract class ProjectController {
```
Comment (Ducatel) [inline BackEnd/src/controllers/ProjectController.php]: Should not be abstract and should not contains only static methods.
You should create an instance and use the dependency injector of slim

File: BackEnd/src/controllers/ProjectController.php [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+    /**
+     * Retrieving all information to list all languages from all versions from all projects stored
```
Comment (Ducatel) [inline BackEnd/src/controllers/ProjectController.php]: Comment not complete

File: BackEnd/src/controllers/ProjectController.php [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+            $projects[] = $project;
+
+            $projectStructure = [];
+        }
+
+        return $projects;
+    }
+
+    /**
+     * @param SplFileInfo $projectFolder
```
Comment (Ducatel) [inline BackEnd/src/controllers/ProjectController.php]: your comment should be usefull.
You must correct it everywhere

File: BackEnd/src/controllers/ProjectController.php [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            $languageStructure = $versionStructure;
+        }
+    }
+
+    /**
+     * Delete doc for project and corresponding backup
+     *
+     * @return null|string null or the error's description if one append
+     */
+    public static function deleteProject($name, $version, $language, $archiveFolder, $storageFolder) {
```
Comment (Ducatel) [inline BackEnd/src/controllers/ProjectController.php]: Should take a Project object

File: BackEnd/src/routes/ProjectRoutes.php [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+        $language = $requestParams['language'];
+    }
+
+    $files = $request->getUploadedFiles();
+    $archive = null;
+
+
+    error_log('Checking provided parameters');
+
+    if ($name === null) {
```
Comment (Ducatel) [inline BackEnd/src/routes/ProjectRoutes.php]: All tests behind should be inside the Project model

File: BackEnd/src/routes/ProjectRoutes.php [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    if (pathinfo($archive->getClientFilename(), PATHINFO_EXTENSION) !== 'zip') {
+        $errorMessage = 'archive is not a zip file';
+    }
+
+    if ($errorMessage !== null) {
+        $response = $response->write($errorMessage);
+        return $response->withStatus(400);
+    }
+
+    error_log('Parameters OK');
```
Comment (Ducatel) [inline BackEnd/src/routes/ProjectRoutes.php]: You should not use that. 
Use PSR-3 compatible logger (like [monolog](https://github.com/Seldaek/monolog) for example )

**Pull request merged**